x-pack/plugins/infra/public/containers/logs/log_entries/index.ts

```diff
@@ -124,48 +144,99 @@ const useFetchEntriesEffect = (
+        endTimestamp: overrides.endTimestamp || props.endTimestamp,
+        query: overrides.filterQuery || props.filterQuery || undefined, // FIXME
+      };
+
+      if (props.timeKey) {
+        (fetchArgs as LogEntriesCenteredRequest).center = props.timeKey;
+      } else {
+        (fetchArgs as LogEntriesBeforeRequest).before = 'last';
+      }
+
+      const { data: payload } = await fetchLogEntries(fetchArgs);
       dispatch({ type: Action.ReceiveNewEntries, payload });
+
+      // Move position to the bottom if it's the first load.
+      // Do it in the next tick to allow the `dispatch` to fire
+      if (!props.timeKey && payload.bottomCursor) {
+        setTimeout(() => {
+          props.jumpToTargetPosition(payload.bottomCursor!);
+        });
+      } else if (
+        props.timeKey &&
+        payload.topCursor &&
+        payload.bottomCursor &&
+        !timeKeyIsBetween(payload.topCursor, payload.bottomCursor, props.timeKey)
+      ) {
+        props.jumpToTargetPosition(payload.topCursor);
+      }
     } catch (e) {
       dispatch({ type: Action.ErrorOnNewEntries });
     }
   };
 
   const runFetchMoreEntriesRequest = async (direction: ShouldFetchMoreEntries) => {
-    dispatch({ type: Action.FetchingMoreEntries });
+    if (!props.startTimestamp || !props.endTimestamp) {
+      return;
+    }
     const getEntriesBefore = direction === ShouldFetchMoreEntries.Before;
-    const timeKey = getEntriesBefore
-      ? state.entries[0].key
-      : state.entries[state.entries.length - 1].key;
-    const getMoreLogEntries = getEntriesBefore ? getLogEntriesBefore : getLogEntriesAfter;
+
+    // Control cursors are correct
+    if (getEntriesBefore && !state.topCursor) {
+      return;
+    } else if (!state.bottomCursor) {
+      return;
+    }
+
+    dispatch({ type: Action.FetchingMoreEntries });
+
     try {
-      const payload = await getMoreLogEntries({ ...props, timeKey });
+      const fetchArgs: LogEntriesRequest = {
+        sourceId: props.sourceId,
+        startTimestamp: props.startTimestamp,
+        endTimestamp: props.endTimestamp,
+        query: props.filterQuery || undefined, // FIXME
+      };
+
+      if (getEntriesBefore) {
+        (fetchArgs as LogEntriesBeforeRequest).before = state.topCursor!; // We check for nullity above already
+      } else {
+        (fetchArgs as LogEntriesAfterRequest).after = state.bottomCursor;
+      }
+
+      const { data: payload } = await fetchLogEntries(fetchArgs);
+
       dispatch({
         type: getEntriesBefore ? Action.ReceiveEntriesBefore : Action.ReceiveEntriesAfter,
         payload,
       });
-      return payload.entriesEnd;
+
+      return payload.bottomCursor;
     } catch (e) {
       dispatch({ type: Action.ErrorOnMoreEntries });
     }
   };
 
   const fetchNewEntriesEffectDependencies = Object.values(
-    pick(props, ['sourceId', 'filterQuery', 'timeKey'])
+    pick(props, ['sourceId', 'filterQuery', 'timeKey', 'startTimestamp', 'endTimestamp'])
```

Comment:
I think [this comment](https://github.com/elastic/kibana/pull/54280#discussion_r390588273) applies here as well.

I understand this array as "trigger the effect when these value change", not "these are the values that the effect uses".

Reply:
The latter is the case, though. From the React `useEffect` docs:

> If you use this optimization, make sure the array includes all values from the component scope (such as props and state) that change over time and that are used by the effect. Otherwise, your code will reference stale values from previous renders. Learn more about how to deal with functions and what to do when the array values change too often.
>
> If you want to run an effect and clean it up only once (on mount and unmount), you can pass an empty array ([]) as a second argument. This tells React that your effect doesn’t depend on any values from props or state, so it never needs to re-run. This isn’t handled as a special case — it follows directly from how the dependencies array always works.
>
> If you pass an empty array ([]), the props and state inside the effect will always have their initial values. While passing [] as the second argument is closer to the familiar componentDidMount and componentWillUnmount mental model, there are usually better solutions to avoid re-running effects too often. Also, don’t forget that React defers running useEffect until after the browser has painted, so doing extra work is less of a problem.
>
> We recommend using the exhaustive-deps rule as part of our eslint-plugin-react-hooks package. It warns when dependencies are specified incorrectly and suggests a fix.
>
> The array of dependencies is not passed as arguments to the effect function. Conceptually, though, that’s what they represent: every value referenced inside the effect function should also appear in the dependencies array. In the future, a sufficiently advanced compiler could create this array automatically.  

Dan Abramov's blog post https://overreacted.io/a-complete-guide-to-useeffect/#dont-lie-to-react-about-dependencies also stresses the point.

Reply:
> If you use this optimization, make sure the array includes all values from the component scope (such as props and state) that change over time and that are used by the effect. Otherwise, your code will reference stale values from previous renders.

This is misleading. I get why they say it but it's incorrect. As long as one dependency changes, the callback/effect will close again and get the latest values.

See https://codesandbox.io/s/hooks-deps-playground-uphs8, in particular this callback:

```ts
const countPlusOne = useCallback(() => setCount(count + 1), [color]);
```

If the statement above were true, calling `countPlusOne()` will always return 1. The value of `count` when the component renders is `0`. The function would close over that value and always return 1. The dependencies don't include `count`, only `color`.  If that statement would be true, the value of `count` would never update.

`color` gets updated once per second. When its value changes, `useCallback` refreshes the function it points to. Now `count` refreshes its value, even if it's not in the dependency list, and the counter can be incremented again, when `color` changes.

![deps-changing-demo](https://user-images.githubusercontent.com/57448/76419099-17985d80-63a0-11ea-83a5-0241d6df1313.gif)

This example uses `useCallback`, but the same holds true for `useEffect`. In that case, the effect will run when any value of the dependency array changes.

This dependency will always run and will always have the latest values, even if doesn't specify any value in the dependency array.
```ts
function SomeComponent() {
  const [count, setCount] = useState(0);
  const [color, setColor] = useState('#fff');

  useEffect(() => { console.log({ color, count }) }, [Math.random()]); // Neither `color` nor `count` are here.
}
```

---

Back to our code, the suggestion you are making will trigger `useEffect` in situations where it shouldn't be called. The change needed will be deeper than just changing the dependency list. We will need to change the callbacks themselves to make sure they have early returns, something that the current dependency lists gives us for free.

Reply:
I think I understand your concern, from the Dan Abramov's quote:

> The array of dependencies is not passed as arguments to the effect function. Conceptually, though, that’s what they represent

I think our disagreement comes from that sentence. Although that's what the dependencies _represent_ is not what they _are_, which is how I'm using them, but I think your concern comes from it: that way of using them is seen as hacky.

I want to address your concern but I think to do so with the current state will delay this PR even more, [and we want to change the whole hook anyway](https://github.com/elastic/kibana/pull/54280?email_token=AAAOA2CLUPDUFPCJQSIAPT3RG5SA5A5CNFSM4KELMYD2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCY275NY#discussion_r390437924).

I personally think it's better to move forward as it is and address it when we refactor the hook.

What do you think?

Reply:
I agree, let's get this PR in and follow it up with a fix soon after.